distod/src/main/scala/com/github/codelionx/util/trie/CandidateTrie.scala

```diff
@@ -0,0 +1,146 @@
+package com.github.codelionx.util.trie
+
+import com.github.codelionx.distod.types.CandidateSet
+
+import scala.annotation.tailrec
+import scala.collection.{mutable, Factory, StrictOptimizedIterableOps}
+import scala.language.implicitConversions
+
+
+class CandidateTrie[V]
+  extends mutable.Map[CandidateTrie.K, V]
+    with mutable.MapOps[CandidateTrie.K, V, mutable.Map, CandidateTrie[V]]
+    with StrictOptimizedIterableOps[(CandidateTrie.K, V), mutable.Iterable, CandidateTrie[V]] {
+
+  import CandidateTrie._
+
+
+  private val suffixes: mutable.Map[A, CandidateTrie[V]] = mutable.Map.empty
+  private var value: Option[V] = None
+
+  override def get(key: K): Option[V] =
+    if (key.isEmpty) value
+    else
+    // key.head is safe here
+      suffixes.get(key.head).flatMap(_.get(key.tail))
+
+  @tailrec
+  final def withPrefix(prefix: K): CandidateTrie[V] =
+    if (prefix.isEmpty) this
+    else {
+      // prefix.head is safe here
+      val head = prefix.head
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Usage of head on collections.](https://app.codacy.com/app/CodeLionX/distod/pullRequest?prid=4697513)